### PR TITLE
[WIP] Trainer supports all Datasets as train_dataset, with/without __len__ #5990

### PR DESCRIPTION
**Pull request by j-rossi-nl**
A first PR.

Passed:
- `make test`
- `make style`
- `make quality`

Modifies:
- trainer.py: fixes issue
- test_trainer.py: calls `Trainer.train()`

It fixes only the case where the TRAINING dataset has not the `__len__` method.

The distinction is not between `Dataset` and `IterableDataset`, but between objects that are instances of class where `__len__` is implemented or not. This is pointed in [pytorch source](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#Dataset), implementation of `__len__` is up to the user.
The test is therefore: `isinstance(dataset, collections.Sized)`

NB: fixing for EVAL and TEST dataset will require more code refactor: all get funneled to `Trainer._prediction_loop()` without keeping track of whether it is EVAL or TEST, which makes the usage of `TrainingArguments.eval_steps` impossible to assume. (not to mention, there is no `test_steps` field in `TrainingArguments`)

Not all datasets have an implementation of `__len__` method, therefore the trainer should not assume it is available.

The use case is:
- Dataset with `__len__`: use `num_train_epochs` or `max_steps` to specify how long should training run
- Dataset without `__len__`: use only `max_steps`

The limitation is still valid on the EVAL / TEST dataset who still has to implement `__len__`

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -225,12 +226,19 @@ def __init__(
...
+            raise ValueError("train_dataset does not implement __len__, max_steps has to be specified")
+        if eval_dataset is not None and not isinstance(eval_dataset, collections.Sized):
+            raise ValueError("Eval Dataset: eval_dataset must implement __len__")
 
     def get_train_dataloader(self) -> DataLoader:
         """
         Returns the training :class:`~torch.utils.data.DataLoader`.
         """
-        if isinstance(self.train_dataset, torch.utils.data.IterableDataset):
+        if isinstance(self.train_dataset, collections.Sized):
```
Comment (sgugger) [inline src/transformers/trainer.py:241]: I think you mean not here and below.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR. Note that there is some moving around of the code in Trainer coming in #5982 (I'll probably merge it today) so you may need to adapt a bit the code.
Note that there is no test_steps field since the evaluation is supposed to be complete on the test set (users can always pass along a shorter dataset if they want).

File: src/transformers/trainer.py (hunk 1 above)
Comment (j-rossi-nl) [reply src/transformers/trainer.py]: Exactly, I fixed it in commit 8dd9ae0.

**Comment by j-rossi-nl** [comment]
> Note that there is some moving around of the code in Trainer coming in #5982 (I'll probably merge it today) so you may need to adapt a bit the code.

I'll wait for the merge of #5982 and introduce the fix for #5990 after.

> Note that there is no test_steps field since the evaluation is supposed to be complete on the test set (users can always pass along a shorter dataset if they want).

I see. 
At the moment, the functionality is that it will refuse a dataset that does not implement `__len__`, whether it is the EVAL dataset or the TEST dataset.

**Comment by j-rossi-nl** [comment]
Hi,

- Merged to include all changes from #5982
- Now accepts `IterableDataset` for any dataset (TRAIN, EVAL, TEST)
- Displays information (how many steps, etc...)
- `test_trainer.py` includes an end-to-end test of `train()` and `predict()`

It fixes the issue #5990 with my code.

Checklist:
- `make test` is positive (all passed, no failed)
- `make style` and `make quality`

Looking forward to review.

I'm still unsure about how to test what the dataset is:
- the  type hinting says `train_dataset: Dataset`
- `pytorch` indicates it is good practice to implement `__len__` on Map-Style dataset, but in the code there is no way this could be enforced
- `pytorch` relies only on the logic: user will inherit `Dataset` and implement `__len__` or user will inherit `IterableDataset` and not implement `__len__`. In `Dataloader` every time there is a doubt it is checking if the object is an instance of `Dataset` or `IterableDataset`
- in my code, I followed my first rule: either the object has `__len__` or it does not. 

Still wondering if I should code it `pytorch`-style and trust the given answer blindly, or keep it the way it is done, which is a bit more paranoid ?
Any comment appreciated.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -225,9 +226,16 @@ def __init__(
                 ),
                 FutureWarning,
             )
+        # Some warning for the user related to the usage of datasets that do not implement __len__
+        if args.max_steps > 0:
+            logger.warning("max_steps is given, it will override any value given in num_train_epochs")
+        if not isinstance(train_dataset, collections.abc.Sized) and args.max_steps == 0:
```
Comment (sgugger) [inline src/transformers/trainer.py:232]: You should check there is a `train_dataset` by adding `train_dataset is not None` in your test (the Trainer could be used just for eval/predictions).

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -880,8 +927,9 @@ def prediction_loop(
         # inside a DistributedDataParallel as we'll be under `no_grad` anyways.
 
         batch_size = dataloader.batch_size
+        num_examples = self.num_examples(dataloader) if is_dataset_sized else self.args.eval_steps * batch_size
```
Comment (sgugger) [inline src/transformers/trainer.py:930]: `self.args.eval_steps` is the number of training steps between two evaluation phases, so it can't be used here for that purpose. Since there is an error when `eval_dataset` is not sized, there should not be any test here.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -260,7 +268,7 @@ def get_train_dataloader(self) -> DataLoader:
         )
 
     def _get_eval_sampler(self, eval_dataset: Dataset) -> Optional[torch.utils.data.sampler.Sampler]:
-        if isinstance(eval_dataset, torch.utils.data.IterableDataset):
+        if not isinstance(eval_dataset, collections.abc.Sized):
```
Comment (sgugger) [inline src/transformers/trainer.py:271]: This case should use the same defensive test as in init: if the user provides a non-sized evaluation/test set, there should be a ValueError here as the rest won't work.

**Review by sgugger — CHANGES_REQUESTED** [review]
Thanks a lot for taking care of the merge! There were still some conflicts, but I dealt with them.

I like the fact that you directly test if the `Dataset` has a length or not. It seems more solid since PyTorch does not check during their `Dataset` init if they actually have a length, so we can't really rely on the distinction `Dataset`/`IterableDataset`.

I just think that since (for now) the choice is to reject "infinite" datasets at evaluation/prediction (which seems ok to me since I don't see how this could actually give some results), we should be consistent in the code. I've made a few comments to that effect if you can adapt your PR.

File: src/transformers/trainer.py (hunk 2 above)
Comment (j-rossi-nl) [reply src/transformers/trainer.py]: Done.

File: src/transformers/trainer.py (hunk 4 above)
Comment (j-rossi-nl) [reply src/transformers/trainer.py:271]: Done

**Comment by j-rossi-nl** [comment]
- the test is whether a `Dataset` object has `__len__` or not
- iterable dataset is OK for training,  __only__ if `max_steps` has a strictly positive value
- iterable dataset is not acceptable for evaluation or prediction

The confusion about `eval_steps` has been purged.

The test `test_trainer_iterable_dataset` in `test_trainer.py` has been extended to check for corner cases and associated exceptions. Only the exception type is checked, not exception message.

Looking forward to review.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -818,6 +861,9 @@ def evaluate(self, eval_dataset: Optional[Dataset] = None) -> Dict[str, float]:
         Returns:
             A dictionary containing the evaluation loss and the potential metrics computed from the predictions.
         """
+        if eval_dataset is not None and not isinstance(eval_dataset, collections.abc.Sized):
```
Comment (sgugger) [inline src/transformers/trainer.py:864]: Nit: this will be raised in `self.get_eval_dataloader` so not necessary here.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -849,6 +896,8 @@ def predict(self, test_dataset: Dataset) -> PredictionOutput:
             metrics (:obj:`Dict[str, float]`, `optional`):
                 The potential dictionary of metrics (if the dataset contained labels).
         """
+        if not isinstance(test_dataset, collections.abc.Sized):
```
Comment (sgugger) [inline src/transformers/trainer.py:899]: Same comment as above.

**Review by sgugger — APPROVED** [review]
Just two nit-picking, but this looks great to me, thanks for the work!

**Comment by j-rossi-nl** [comment]
Nitpicking
- removed redundant test on dataset (eval / test)

Checklist:
- `make test` passed (no fail)
- `make style`
- `make quality`

Up for review again.

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -300,13 +308,13 @@ def get_test_dataloader(self, test_dataset: Dataset) -> DataLoader:
...
 
-        Will use no sampler if :obj:`test_dataset` is a :obj:`torch.utils.data.IterableDataset`, a sequential
+        Will use no sampler if :obj:`test_dataset` does not implement :obj:`__len__`, a sequential
         sampler (adapted to distributed training if necessary) otherwise.
 
         Subclass and override this method if you want to inject some custom behavior.
 
         Args:
-            eval_dataset (:obj:`torch.utils.data.dataset.Dataset`, `optional`):
+            test_dataset (:obj:`torch.utils.data.dataset.Dataset`, `optional`):
```
Comment (LysandreJik) [inline src/transformers/trainer.py:317]: Why the name change? The variable in the signature is still `eval_dataset`, which is coherent with the signature name `get_eval_dataloader`

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -495,17 +536,16 @@ def train(self, model_path: Optional[str] = None):
...
 
             if is_torch_tpu_available():
-                parallel_loader = pl.ParallelLoader(train_dataloader, [self.args.device]).per_device_loader(
+                train_dataloader = pl.ParallelLoader(train_dataloader, [self.args.device]).per_device_loader(
                     self.args.device
                 )
-                epoch_iterator = tqdm(parallel_loader, desc="Iteration", disable=not self.is_local_master())
-            else:
-                epoch_iterator = tqdm(train_dataloader, desc="Iteration", disable=not self.is_local_master())
+            epoch_iterator = tqdm(train_dataloader, desc="Iteration", disable=not self.is_local_master())
```
Comment (LysandreJik) [inline src/transformers/trainer.py:542]: Nice, cleaner!

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -809,6 +849,9 @@ def evaluate(self, eval_dataset: Optional[Dataset] = None) -> Dict[str, float]:
 
         The calling script will be responsible for providing a method to compute metrics, as they are
         task-dependent (pass it to the init :obj:`compute_metrics` argument).
+        :obj:`eval_dataset` must implement method :obj:`__len__`.
+
+        You can also subclass and override this method to inject custom behavior.
 
         You can also subclass and override this method to inject custom behavior.
```
Comment (LysandreJik) [inline src/transformers/trainer.py:856]: Duplicate

File: tests/test_trainer.py [hunk 10]
```diff
@@ -194,10 +198,50 @@ def test_trainer_eval_lm(self):
...
-        training_args = TrainingArguments(output_dir="./examples", no_cuda=True)
-        trainer = Trainer(model=model, args=training_args, train_dataset=train_dataset)
+        model = AutoModelForMaskedLM.from_pretrained(MODEL_ID)
+        tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
+        train_dataset = SampleIterableDataset(file_path=PATH_SAMPLE_TEXT, tokenizer=tokenizer)
+        training_args = TrainingArguments(output_dir="./examples", no_cuda=True, max_steps=2)
+        data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=True, mlm_probability=0.15)
+
+        # 5990 iterable dataset = no __len__ method
+        # 5990 - Train with iterable dataset, no __len__
```
Comment (LysandreJik) [inline tests/test_trainer.py:211]: Maybe not necessary to add the references to the PRs in the code?

**Review by LysandreJik — APPROVED** [review]
Great!

**Comment by carson-sestili** [comment]
Hi, is this PR still being reviewed? I would like to use `Trainer` with an `IterableDataset` and this looks like exactly what's needed to make that happen. If you have time, I would greatly appreciate this PR to get into the next version :) thank you!

**Comment by j-rossi-nl** [comment]
I realize a part has been merged, but not everything.

**Comment by j-rossi-nl** [comment]
And I can't seem to find a way to re-open this PR. So I guess, I should open a new one, and link to this one...

File: src/transformers/trainer.py (hunk 7 above)
Comment (j-rossi-nl) [reply src/transformers/trainer.py:317]: There are 2 functions:
- `get_eval_dataloader` has argument `eval_dataset` (L280-L305)
- `get_test_dataloader` has argument `test_dataset` (L307-L329)

The name did not change, but the diff makes it look like it changed indeed.

File: src/transformers/trainer.py (hunk 9 above)
Comment (j-rossi-nl) [reply src/transformers/trainer.py:856]: removed duplicate

File: tests/test_trainer.py (hunk 10 above)
Comment (j-rossi-nl) [reply tests/test_trainer.py:211]: It's reference to the Issue.
I think I saw other ones, so I mimicked.

**Comment by j-rossi-nl** [comment]
@carson-sestili The PR #7858 has been merged to master and fixes the bug.
You can already use it by installing from source.

**Comment by carson-sestili** [comment]
@j-rossi-nl Thank you very much!
